### DISCUSSION TITLE
How to do angular closeness centrality from a shapefile

### DISCUSSION THREAD

**Discussion by felipetavares80** [pysal/momepy#462] [Q&A]
Hello everyone.

I'm trying to do a angular weighted closeness centrality from a segmented road map.

I'm importing it from osmnx, simplifying the segments by Douglas-Peucker on Qgis, extracting the points and creating a graph for momepy closeness centrality analysis.

I did the graph but it seems that the results do not match the manual calculations when I set the angular option for closeness centrality. It matches for topological but not for angular.

I don't know if I'm doing something wrong on manual calculations or in the coding, but 'll appreciate a help on this.

Thanks in advance!

All the Best.

The Codes, the manual calculations document and the shapefiles is attached.
[angularclosenesspack.zip](https://github.com/pysal/momepy/files/10479608/angularclosenesspack.zip)

**Comment by martinfleis** [comment]
Hi @felipetavares80,

First question - which version of momepy do you have? There was a bug in creation of angular graph recently (fixed in the latest release).

Second - what makes you think it matches topological calculation? See

```py
import geopandas as gpd
import osmnx as ox
import networkx as nx
import momepy

# load Shapefiles as node/edge GeoDataFrames indexed as described in OSMnx docs
gdf_nodes = gpd.read_file('./teste/nodes.shp').set_index('osmid')
gdf_edges = gpd.read_file('./teste/edges.shp').set_index(['u', 'v', 'key'])

# Extract columns x e y coordinates from points
gdf_nodes['x'] = gdf_nodes.geometry.x
gdf_nodes['y'] = gdf_nodes.geometry.y

# convert the node/edge GeoDataFrames to a MultiDiGraph
graph_attrs = {'crs': 'epsg:32625', 'simplified': True}

#G2 = ox.graph_from_gdfs(gdf_nodes, gdf_edges)
G2 = ox.graph_from_gdfs(gdf_nodes, gdf_edges, graph_attrs)



# saving a gdf and letting momepy do the conversion back to the graph.
edges = ox.graph_to_gdfs(G2, nodes=False, edges=True,
                                   node_geometry=False, fill_edge_geometry=True)

dual = momepy.gdf_to_nx(edges, approach='dual')

# measure with angle as a weight
dual = momepy.closeness_centrality(dual, weight='angle')
# measure without weight - topological
dual = momepy.closeness_centrality(dual, name="topological")

dual_gdf = momepy.nx_to_gdf(dual, points=False)

print(dual_gdf[["closeness", "topological"]].head())
```

You can see that the values are different

```
   closeness  topological
0   0.004002     0.077749
1   0.005351     0.083963
2   0.004401     0.076200
3   0.005510     0.089063
4   0.006022     0.086391
```
Which is also shown on plots:

```py
dual_gdf.plot("closeness")
```
![74a50e6b-a49e-4d03-9c38-34eeee44040f](https://user-images.githubusercontent.com/36797143/214042832-05ab9de0-31d8-43d8-919a-6c5439bb399c.png)

```py
dual_gdf.plot("topological")
```
![51a2ee30-cd37-4f81-935d-c7f5c50141e4](https://user-images.githubusercontent.com/36797143/214042891-c05a0a5d-caa9-4ad1-87bd-67e5bff31094.png)

It seems to work as expected on my side, so I may need a bit more info to see the issue.

> I did the graph but it seems that the results do not match the manual calculations when I set the angular option for closeness centrality. It matches for topological but not for angular.

How do you do the manual calculation?

**Reply by felipetavares80** [reply]
Thank you very much for the reply @martinfleis !

I did install the momepy last week, so I think it is the latest version.

I would like to ask you to take a look in the text document attached, where it is a tiny map and I did the manual calculations and the angular result  doesn't matches. Although the topological does match. The code for this tiny map is also in the zip package.

Also I have tested the length weight and it has the same result as the topological. Something it seems odd for me.

I'll appreciate your comments on this.

Thanks in advance.

All the Best.

**Reply by martinfleis** [reply]
ugh, we have another bug in there.... these are angles reported by momepy. Some are correct but other are inverse.

![2312c81e-37e5-42ba-b707-fabf5a04e522](https://user-images.githubusercontent.com/36797143/214062069-4e26ed39-4e73-4734-9bf6-b8d9b8d72d67.png)

I'm working on a fix. Thanks a lot for flagging it!

**Reply by martinfleis** [reply]
The reason for this is a formula that ignores directionality https://github.com/pysal/momepy/blob/756d9f2c1829f27379f504f5f740dd4b559a7015/momepy/utils.py#L41-L48

**Reply by felipetavares80** [reply]
Oh! Very good!

Looking forward to have the fixed version! Thanks a lot!

All The best.

**Reply by martinfleis** [reply]
Actually, I am not sure about that now. What is shown on the figure above is the weight (what you do manually in you'd doc as 180 - x) that is expected in the computation, so that should be correct I believe.
<img width="753" alt="Screenshot 2023-01-23 at 15 50 00" src="https://user-images.githubusercontent.com/36797143/214069630-3d42ec09-ad72-454a-8914-491de587bc8d.png">


The reason for the difference is the computation of shortest paths. In the diagram above, the shortest path on the graph weighted by angle between points 1 and 3 is not a direct one (that would cost 101) but through node 2. Then the route 1-2-3 costs 1.56 + 80.21 = 81.77. While being mathematically correct, it does not make sense in terms of real-life routing. I need to think about the consequences and how to adjust that... @jGaboardi any ideas?

**Reply by jGaboardi** [reply]
This is an interesting problem. As you say @martinfleis, angular closeness is a highly abstract concept that may not easily be adapted to real-life routing. Since it seems to be a mathematically sound result, should we not keep the current implementation? We can add a note on this caveat in the user guide using this exact example.

**Reply by martinfleis** [reply]
We should probably keep it as is.

 @songololo how is cityseer dealing with this kind of a situation?

**Reply by songololo** [reply]
Hi All, yes, this is a known complication of simplest-path (angular) centralities - I can't remember exactly which paper but it was discussed by Turner, I think in his 2007 paper. See Figure 3 in [linked working paper](https://arxiv.org/pdf/2106.14040.pdf) for a diagram, and see the references for the Turner paper info. [EDIT - clarified wording]

The way to handle this is to split the bidirectional network into a directed network (per Figure 3 in the working paper) which enforces onwards-only movement (i.e. no backtracking) or to check in the shortest path algorithm that the heuristic is not backtracking to reach a node that could have been reached directly from a prior node. Cityseer does the latter approach by keeping track of previously visited nodes per [this line](https://github.com/benchmark-urbanism/cityseer-api/blob/3503c51f10944e057d71814254b73781ab5aa4a6/cityseer/algos/centrality.py#L103).

**Reply by martinfleis** [reply]
It may actually be the best to depend on `cityseer` directly in this. There's no need to re-implement the algorithm and maintain it at two places. I'm just reluctant to expose yet another object (`cityseer.structures.NetworkStructure`) momepy users need to deal with... It would certainly break the logic we have now, of having a pure networkx graph for network analysis and GeoDataFrame for geometry analysis. 

@songololo can you think about a shortcut converting a networkx object created by `momepy.nx_from_gdf` to a structure needed by city seer and retrieving the values from it so we can attach them as node/edge attributes? That way the signature of momepy's centrality measures does not have to change and cityseer object would not be exposed here.

**Reply by songololo** [reply]
Will try to look at this during the next week or so.

**Reply by songololo** [reply]
For discussion, here is how this would roughly look via `cityseer`. I'm not using `osmnx` for conversion to a GeoDataframe because it seems to use directed networks, which are not compatible with `cityseer`. Once computed the columns could be copied back and forth between dataframes?

```python
import geopandas as gpd
import networkx as nx
from cityseer.metrics import networks
from cityseer.tools import graphs, plot

# cityseer uses undirected networks - assumes pedestrians
G1 = nx.MultiGraph()
# add nodes
gdf_nodes = gpd.read_file("./temp/nodes.shp").set_index("osmid")
for idx, row in gdf_nodes.iterrows():
    G1.add_node(idx, x=row["geometry"].x, y=row["geometry"].y)
# add edges
gdf_edges = gpd.read_file("./temp/edges.shp").set_index(["u", "v", "key"])
for (idx_a, idx_b, edge_idx), row in gdf_edges.iterrows():
    G1.add_edge(idx_a, idx_b, edge_idx, geom=row["geometry"])
# visual check
plot.plot_nx(G1, labels=True)

# calculating angular centrality over the primal representation
nodes_gdf, network_structure = graphs.network_structure_from_nx(G1, crs=32625)
# network structure contains information for jitted functions / numpy like
nodes_gdf = networks.node_centrality(
    measures=["node_harmonic_angular"],
    network_structure=network_structure,
    nodes_gdf=nodes_gdf,
    distances=[100, 400],
    angular=True,
)
print(nodes_gdf.head())

# using the dual representation
G1_dual = graphs.nx_to_dual(G1)
# use plot_geoms to show geometric representation
plot.plot_nx(G1_dual, labels=True, plot_geoms=True)
nodes_gdf_dual, network_structure_dual = graphs.network_structure_from_nx(
    G1_dual, crs=32625
)
nodes_gdf_dual = networks.node_centrality(
    measures=["node_harmonic_angular"],
    network_structure=network_structure_dual,
    nodes_gdf=nodes_gdf_dual,
    distances=[100, 400],
    angular=True,
)
print(nodes_gdf_dual.head())
```

**Reply by songololo** [reply]
It would be possible to create a convenience method for going directly from `momepy` GeoDataframes to cityseer data structures and back again. i.e. to skip the `networkX` step.

**Reply by songololo** [reply]
For info - the `network_structure` in `cityseer` is basically a data structure containing numpy arrays which are used by the jitted functions (so that it can scale to larger graphs). EDIT: This also pre-computes information such as in-out bearings for segments and segment impedances (distance and angular change), etc.
